### PR TITLE
Add sapphire optimizers, refactor core gradient solvers

### PR DESCRIPTION
**Pull request by zjf4**
Adds Sapphire optimizers + refactors proximal gradient and Sapphire to rely on the same underlying gradient solver framework.

### CODE REVIEW

File: rlaopt/atoms/linear_model/_base.py [hunk 1]
```diff
@@ -111,6 +111,17 @@ def forward(self):
 
         return self._get_loss(beta_tensor, intercept_tensor)
 
+    def evaluate(self, variable_values):
+        if self.dataloader._shuffle:
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_base.py]: Why are we accessing a private attribute? Let's make sure this actually makes sense to do.

File: rlaopt/solvers/gradient_solvers/gradient_solver_configs.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+class ProxGradConfig(GradSolverConfig):
+    """Configuration for proximal gradient solvers."""
+
+    eta: float = Field(
+        default=1.0, gt=0.0, description="Step size for the gradient update."
+    )
+
+    use_acceleration: bool = Field(
+        default=False, description="Whether to use acceleration techniques."
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_configs.py]: Let's change this description to "Whether to use Nesterov acceleration." If it's not Nesterov acceleration, we can just say "Whether to use acceleration."

File: rlaopt/solvers/gradient_solvers/gradient_solver_configs.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    """Configuration for Sapphire solver."""
+
+    # Base optimizer specification.
+    base_method: Literal["saga", "svrg", "sgd"] = Field(
+        default="saga", description="Base gradient method to use in Sapphire."
+    )
+
+    # Stepsize selection
+    eta: float = Field(
+        default=0.10,
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_configs.py]: Shorten from 0.10 to 0.1

File: rlaopt/solvers/gradient_solvers/gradient_solver_configs.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    precond_update_freq: int = Field(
+        default=2,
+        gt=0,
+        description="How frequently in epochs the preconditioner and "
+        "stepsize (if auto_update_stepize = True) are updated. "
+        "Defaults to 2 epochs.",
+    )
+
+    subproblem_iters: int = Field(
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_configs.py:77]: I'm assuming this is not actually used if the problem is fully smooth right? We might want to add this to the description.

File: rlaopt/solvers/gradient_solvers/gradient_solver_core.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+error metrics used by all gradient based solvers.
+"""
+
+from abc import ABC, abstractmethod
+from dataclasses import replace
+from functools import partial
+from typing import Any, Callable
+
+import torch
+from torch.nn.modules.loss import _Loss
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_core.py]: Any particular reason to use the internal class `_Loss` from `torch`?

File: rlaopt/solvers/gradient_solvers/gradient_solver_states.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+from dataclasses import dataclass
+
+import torch
+
+from rlaopt.ext_tensordict import TensorDict
+from rlaopt.linalg.preconditioners.identity import Identity, IdentityConfig
+from rlaopt.linalg.preconditioners.nystrom import Nystrom
+from rlaopt.solvers.solver_base import SolverState
+from rlaopt.splitting.prox_grad_split import ProxGradSplit
+from rlaopt.splitting.sapphire_split import SapphireSplit
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_states.py]: Let's import these directly from `rlaopt.splitting`

File: rlaopt/solvers/gradient_solvers/gradient_solver_states.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    and accelerated stochastic methods.
+
+    Attributes:
+        P: Preconditioner matrix (Nystrom approximation or Identity).
+            Defaults to Identity.
+        precond_update_freq: Frequency (in iterations) for updating the
+            preconditioner. Defaults to 1.
+    """
+
+    P: Nystrom | Identity = Identity(IdentityConfig())
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/gradient_solver_states.py]: Do we need to make `P` have a default initialization, given that we already assign it?

Also I would prefer to not import `Nystrom` or `Identity` here. Let's use the `Preconditioner` class instead as the typehint.

File: rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    Aop: Callable[[torch.Tensor], torch.Tensor] | LinearOperator,
+    shape: tuple[int, int],
+    device: torch.device,
+) -> SapphireState:
+    L = randomized_powering(Aop, shape, device=device)
+
+    if isinstance(Aop, _SubampHVPLinOp):
+        L += 1e-2
+    else:
+        L = 2 * L
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py]: Why add 1e-2 or multiply by 2? I vaguely remember having a fudge factor for one of SVRG/SAGA. But I don't quite get adding 1e-2.

File: rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    dtype: torch.dtype,
+    update_stepsize: bool = True,
+) -> Callable[[TensorDict, SapphireState], SapphireState]:
+    """Build preconditioner update function.
+
+    Returns:
+        Callable for updating the preconditioner
+    """
+    preconditioner_class = get_preconditioner_class(precond_config)
+    P = preconditioner_class(precond_config)
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py]: I believe we can just use `get_preconditioner` from the factory directly if we compute the linop first, no? That way `get_preconditioner_class` can have its name changed back to `_get_preconditioner_class` and remain an implementation detail.

File: rlaopt/splitting/linops.py [hunk 10]
```diff
@@ -86,12 +89,88 @@ def __init__(self, smooth_expr: Expression, variable_values: TensorDict):
 
     def _matmul_impl(self, v: torch.Tensor) -> torch.Tensor:
         """Compute Hessian @ v using forward-over-reverse autodiff."""
+        return _pearlmutter_hvp(
+            lambda x: self._smooth_expr.evaluate(x), self._variable_values, v
+        )
+
+
+class _SubampHVPLinOp(LinearOperator):
```
Comment (pratikrathore8) [inline rlaopt/splitting/linops.py]: Rename to `_SubsampHVPLinOp` throughout the codebase, please!

File: rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+def _update_stepsize(
+    state: SapphireState,
+    Aop: Callable[[torch.Tensor], torch.Tensor] | LinearOperator,
+    shape: tuple[int, int],
+    device: torch.device,
+) -> SapphireState:
+    L = randomized_powering(Aop, shape, device=device)
+
+    if isinstance(Aop, _SubampHVPLinOp):
```
Comment (pratikrathore8) [inline rlaopt/solvers/gradient_solvers/precond_update_fn_builder.py]: I don't love that we're using an internal class from a different module here. Is there any way around this?

File: rlaopt/splitting/sapphire_split.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
+"""SapphireSplit class for representing finite-sum composite objective functions."""
+
+import torch
+
+from rlaopt.atoms import Atom
+from rlaopt.atoms.linear_model.linear_model import LinearModel
+from rlaopt.data import DataLoader
+from rlaopt.expression import AddExpression, Expression
+from rlaopt.ext_tensordict import TensorDict
+from rlaopt.splitting.linops import _SubampHVPLinOp
```
Comment (pratikrathore8) [inline rlaopt/splitting/sapphire_split.py]: Again, let's try not to use a "private" class from another module

File: rlaopt/splitting/sapphire_split.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
+"""SapphireSplit class for representing finite-sum composite objective functions."""
+
+import torch
+
+from rlaopt.atoms import Atom
+from rlaopt.atoms.linear_model.linear_model import LinearModel
```
Comment (pratikrathore8) [inline rlaopt/splitting/sapphire_split.py]: Maybe we should just add `LinearModel` to the `rlaopt.atoms` __init__.py file to avoid this import going so deep into the atoms module.

File: rlaopt/splitting/sapphire_split.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+
+        Args:
+            beta_value (TensorDict): Linear model weights at which to evaluate
+                the Hessian.
+            X_batch (torch.Tensor): Batch of feature data for subsampling.
+            y_batch (torch.Tensor): Batch of target labels for subsampling.
+            device (torch.device): Device on which the model parameters live.
+
+        Returns:
+            _SubampHessianLinOp: A linear operator representing the subsampled Hessian.
```
Comment (pratikrathore8) [inline rlaopt/splitting/sapphire_split.py]: Return type is out of date here

File: rlaopt/demo.ipynb [hunk 15]
```diff
@@ -0,0 +1,333 @@
+{
```
Comment (pratikrathore8) [inline rlaopt/demo.ipynb]: Let's delete this notebook!

File: tests/solvers/test_prox_grad.py [hunk 16]
```diff
@@ -200,18 +201,50 @@ def test_complex_mixed_variables(
...
+                stopping_criteria=GradSolverStoppingCriteria(max_iters=500)
+            )
+            x_sol = result.variable_values["x"]
+            return 1 / n_test * torch.linalg.norm(A_test @ x_sol - b_test) ** 2
+
+        mu = torch.tensor(0.1, dtype=torch.float32)
+        grad, _ = torch.func.grad_and_value(test_loss)(mu)
+
+        assert torch.isfinite(grad), f"Gradient w.r.t. mu should be finite, got {grad}"
+        assert grad != 0.0, "Gradient w.r.t. mu should be non-zero"
```
Comment (pratikrathore8) [inline tests/solvers/test_prox_grad.py:247]: Is this enough for the test? I think we should check the gradient value matches what we would expect

File: tests/solvers/test_sapphire.py [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+                -1
+                / X_test.shape[0]
+                * torch.sum(y_test * log_p + (1 - y_test) * (-logits + log_p))
+            )
+
+        mu = torch.tensor(0.01, dtype=precision)
+        grad, _ = torch.func.grad_and_value(test_loss)(mu)
+
+        assert torch.isfinite(grad), f"Gradient w.r.t. mu should be finite, got {grad}"
+        assert grad != 0.0, "Gradient w.r.t. mu should be non-zero"
```
Comment (pratikrathore8) [inline tests/solvers/test_sapphire.py:423]: Same here, is there any way we can check against the actual value we would expect for the gradient?

**Review by pratikrathore8 — CHANGES_REQUESTED** [review]
Hey, it looks good for the most part. Take a look at my comments and let me know what you think.

**Review by pratikrathore8 — APPROVED** [review]
Thanks for making these changes. Looks good to me
